src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -557,33 +558,69 @@ private HttpHandlerDiagnosticListener() : base(DiagnosticListenerName)
 
         private void RaiseRequestEvent(HttpWebRequest request)
         {
-            // If System.Net.Http.Request is on, raise the event
-            if (this.IsEnabled(RequestWriteName))
+            if (request.Headers.Get(RequestIdHeaderName) != null)
             {
-                long timestamp = Stopwatch.GetTimestamp();
-                this.Write(RequestWriteName,
-                    new
+                // this request was instrumented by previous RaiseRequestEvent
+                return;
+            }
+
+            if (this.IsEnabled(ActivityName, request))
+            {
+                var activity = new Activity(ActivityName);
+
+                // Only send start event to users who subscribed for it, but start activity anyway
+                if (this.IsEnabled(RequestStartName))
+                {
+                    long timestamp = Stopwatch.GetTimestamp();
+                    this.StartActivity(activity,
+                        new
+                        {
+                            Request = request,
+                            Timestamp = timestamp
+                        });
+                }
+                else
+                {
+                    activity.Start();
+                }
+
+                request.Headers.Add(RequestIdHeaderName, activity.Id);
+                //we expect baggage to be empty or contain a few items
+                using (IEnumerator<KeyValuePair<string, string>> e = activity.Baggage.GetEnumerator())
+                {
+                    if (e.MoveNext())
                     {
-                        Request = request,
-                        Timestamp = timestamp
+                        StringBuilder baggage = new StringBuilder();
+                        do
+                        {
+                            KeyValuePair<string, string> item = e.Current;
+                            baggage.Append(item.Key).Append('=').Append(item.Value).Append(',');
+                        }
+                        while (e.MoveNext());
+                        baggage.Remove(baggage.Length - 1, 1);
+                        request.Headers.Add(CorrelationContextHeaderName, baggage.ToString());
                     }
-                );
+                }
+
+                // There is no gurantee that Activity.Current will flow to the Response, so let's stop it here
+                activity.Stop();
             }
         }
 
         private void RaiseResponseEvent(HttpWebRequest request, HttpWebResponse response)
         {
-            if (this.IsEnabled(ResponseWriteName))
+            // Response event could be received several times for the same request
+            if (request.Headers[RequestIdHeaderName] != null)
```

Comment:
Note that there is no reliable way to suppress additional `System.Net.Http.Desktop.HttpRequestOut.Stop` events for redirected requests, but there is just one `System.Net.Http.Desktop.HttpRequestOut.Start` event.